src/charmed_kubeflow_chisme/status_handling/multistatus.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+# From prototype for stateless multi-status handling
+# https://github.com/benhoyt/test-charms/blob/statustest-stateless/statustest/src/multistatus.py
+"""Status prioritiser."""
+
+import logging
+import typing
+from typing import List, Optional, Tuple
+
+import ops
+from ops import EventBase, Framework, Unit
+
+logger = logging.getLogger(__name__)
+
+
+class Prioritiser:
+    """Status prioritiser: track the highest-priority status among several components."""
+
+    _PRIORITIES = {
+        "error": 0,
+        "blocked": 1,
+        "waiting": 2,
+        "maintenance": 3,
+        "active": 4,
+        "unknown": 5,
+    }
+
+    def __init__(self):
+        self._components = {}
+
+    def add(self, component: str, get_status: typing.Callable[[], ops.StatusBase]):
+        """Add a named status component."""
+        if component in self._components:
+            raise ValueError(f"duplicate component {component!r}")
+        self._components[component] = get_status
+
+    def highest(
+        self, statuses: Optional[List[Tuple[str, ops.StatusBase]]] = None
+    ) -> ops.StatusBase:
+        """Return highest-priority status with message prefixed with component name."""
```

Comment:
Will we be able to see all other status/error/info messages in the log?

Reply:
Not from `highest`, but the CharmReconciler.reconcile calls `all` first so it has all the statuses.  The `highest` goes to the actual charm status on `juju status`, but everything else is in `debug-log`